### ISSUE TITLE
DualSense

### ISSUE THREAD

**Issue by fabrecordz64** [jfedor2/paaas#1]
Hello,

Thanks for sharing the project. Can you help me with replacing the Razer Raion with an actual DS4 or Dualsense controller?
I tried changing the PID/VID but that didn't do it. Since I do not have a Razer to test, I must change the code.

Thanks.

**Comment by jfedor2** [comment]
It won't work with a DualSense. It has to be a licensed PS4 arcade stick or steering wheel or something like that.

**Comment by ohmyboroda** [comment]
> It won't work with a DualSense. It has to be a licensed PS4 arcade stick or steering wheel or something like that.

@jfedor2 hi! Thanks a lot for your work, can you explain more about why Dualsense won't work?

I’m working on proof of concept to connect an original Dualsense controller to a PS5 using an RP2040. The goal is to enable features like buttons combination, remapping, and turbo while maintaining compatibility with the PS5’s security checks.

Here’s the high-level flow of the idea:
Dualsense (USB/HID) → RP2040 (USB Host) → RP2040 (Logic Processing) → RP2040 (USB Device) → PS5 (USB)

**Dualsense Interface (USB Host Mode):** RP2040 acts as a USB host to capture input data from the Dualsense.
**Processing Logic:** Features like button combinations, remapping, and turbo.
**PS5 Interface (USB Device Mode):** RP2040 emulates a USB HID device to send inputs to the PS5.

The PS5 performs security checks to verify the connected device is an original Dualsense controller. The RP2040 must:
- Pass these security checks by accurately emulating the Dualsense’s USB descriptor and communication protocol.
- Return the verification handshake between the Dualsense and PS5 to avoid detection. Probably with USB Passthrough.
- The basic implementation is clear, there are excellent TinyUSB and PICO-PIO-USB libraries for this.

I have found different community developments in this area, including the excellent GP2040-CE, but I have not found a similar solution.

Current issues on implementing USB Passthrough on RP2040:
- Maybe you know how to implemented USB passthrough on RP2040 to handle bidirectional communication between Dualsense and PS5 via USB?
- Are there libraries or examples for USB host/device mode with USB passthrough that could help?
- Is there a way to intercept and return the verification handshake between the Dualsense and PS5?

Any advice, resources, or shared experiences would be greatly appreciated!

**Comment by jfedor2** [comment]
@ohmyboroda The PS5 uses a different authentication scheme. You can't just pass messages back and forth between the console and the controller and expect it to work because each input report the controller sends contains a signature that's only good for the inputs it contains. You change the inputs and the signature is no longer valid and the console will ignore the packet.

**Comment by ohmyboroda** [comment]
@jfedor2 thank you for quick reply. So there is no alternative solution in this case?

**Comment by jfedor2** [comment]
@ohmyboroda You can use an adapter like the Brook Wingman FGC2.

**Comment by ohmyboroda** [comment]
@jfedor2 
> adapter like the Brook Wingman FGC2.

This only works for PS4 games, right?

> You can't just pass messages back and forth between the console and the controller and expect it to work because each input report the controller sends contains a signature that's only good for the inputs it contains. You change the inputs and the signature is no longer valid and the console will ignore the packet.

It is not very clear why Brook authorisation works in such a case, but does not work in the case of Dualsense. Does this use the same USB Passthrough?

**Comment by jfedor2** [comment]
> > adapter like the Brook Wingman FGC2.
> 
> This only works for PS4 games, right?

No, it works for all PS5 games.

> > You can't just pass messages back and forth between the console and the controller and expect it to work because each input report the controller sends contains a signature that's only good for the inputs it contains. You change the inputs and the signature is no longer valid and the console will ignore the packet.
> 
> It is not very clear why Brook authorisation works in such a case, but does not work in the case of Dualsense. Does this use the same USB Passthrough?

No, the FGC2 seems to just have the chip required to do the proper authentication. It doesn't need a licensed controller to work.

**Comment by ohmyboroda** [comment]
I have studied the device. Am I correct in understanding that the connection in this case is as follows?
Brook Wingman FGC2 -> USB1 on RP2040 (Host Passthrough) 
DualSense or any gamepad -> USB2 on RP2040 (Host) -> RP2040 (USB Device) → PS5 (USB)

**Comment by jfedor2** [comment]
No, you just use the FGC2 as an adapter. Whatever -> FGC2 -> PS5.

**Comment by esp32beans** [comment]
I have used a pi pico emulating a generic USB HID gamepad -> FGC2 -> PS5. Yes, it is very easy.

**Comment by ohmyboroda** [comment]
@jfedor2 
> No, you just use the FGC2 as an adapter. Whatever -> FGC2 -> PS5.

I got it, that is, in fact, similar to P5Mate.

Thank you very much for your feedback.

Can you tell me the community or discord where I can, for example, find people who can share their development experience on rp2040?
